File: server2/Services/UserServices/Controllers/UserController.cs
using Microsoft.AspNetCore.Mvc;
using server2.Services.UserServices.Models;
using server2.Services.UserServices.Ports;

namespace server2.Services.UserServices.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class UserController : ControllerBase
    {
        private readonly ILogger<UserController> _logger;
        private readonly IUserService _userService;
        public UserController(
            ILogger<UserController> logger,
            IUserService userService)
        {
            _logger = logger;
            _userService = userService;
        }

        [HttpGet]
        public async Task<ActionResult<IEnumerable<User>>> Get()
        {
            try
            {
                IEnumerable<User>? users = await _userService.GetUsers();
            
                if (users == null || !users.Any())
                {
                    return NotFound();
                }

                return Ok(users);
            }
            catch (Exception e)
            {
                return StatusCode(StatusCodes.Status500InternalServerError, $"Error: {e.Message}");
            }
        }

        [HttpGet("{userId:int}")]
        public async Task<ActionResult<User>> Get(int userId)
        {
            try
            {
                User? userToGet = await _userService.GetUser(userId);

                if (userToGet == null)
                {
                    return NotFound();
                }
    
                return userToGet;
            }
            catch (Exception e)
            {
                return StatusCode(StatusCodes.Status500InternalServerError, $"Error: {e.Message}");
            }    
        }

        [HttpPost]
        public async Task<ActionResult<User>> Post(User user)
        {
            try
            {
                if (user == null)
                {
                    return BadRequest();
                }

                User? userToAdd = await _userService.AddUser(user);
                
                return CreatedAtAction(
                    nameof(Get), new { id = user.Id }, userToAdd);
            }
            catch (IntegrityException e)
            {
                return Conflict(e.Message);
            }
            catch (Exception e)
            {
                return StatusCode(StatusCodes.Status500InternalServerError, $"Error: {e.Message}");
            }
        }

        [HttpPut("{userId:int}")]
        public async Task<ActionResult<User>> Put(int userId, User user)
        {
            try
            {
                if (user == null || userId != user.Id)
                {
                    return BadRequest();
                }
                
                User? userToUpdate = await _userService.UpdateUser(userId, user);

                if (userToUpdate == null)
                {
                    return NotFound();
                }
                
                return userToUpdate;
            }
            catch (Exception e)
            {
                return StatusCode(StatusCodes.Status500InternalServerError, $"Error: {e.Message}");
            }
        }

        [HttpDelete("{userId:int}")]
        public async Task<ActionResult<User>> Delete(int userId)
        {
            try
            {
                User? userToDelete = await _userService.DeleteUser(userId);

                if (userToDelete == null)
                {
                    return NotFound();
                }
                
                return userToDelete;
            }
            catch (Exception e)
            {
                return StatusCode(StatusCodes.Status500InternalServerError, $"Error: {e.Message}");
            }
        }
    }
}

File: registration-form-express/server/src/controllers/userController.js
const UserService = require('../services/userService');


const UserController = {

  createUser: async (req, res) => {

    try {

      const userData = req.body;

      const newUser = await UserService.createUser(userData);
      res.json(newUser);

    } catch (error) {
  
      res.status(500).json({
        message: error.message
      });
  
    }

  },
  
  getUsers: async (req, res) => {

    try {
      
      const users = await UserService.getUsers();
      res.json(users)
  
    } catch (error) {
      
      res.status(500).json({
        message: error.message
      });
  
    }
    
  },

  getUser: async (req, res) => {

    try {

      const id = req.params.id;
      
      const user = await UserService.getUser(id);
      res.json(user)
      
    } catch (error) {
      
      res.status(500).json({
        message: error.message
      });

    }

  },

  updateUser: async (req, res) => {

    try {
      
      const id = req.params.id;
      const userData = req.body;
      
      const user = await UserService.updateUser(id, userData);
      res.json(user);

    } catch (error) {

      res.status(500).json({
        message: error.message
      });

    }

  },

  deleteUser: async (req, res) => {

    try {
      
      const id = req.params.id;
      
      const user = await UserService.deleteUser(id);
      res.status(200).json({
        message: 'User deleted successfully'
      })

    } catch (error) {
      
      res.status(500).json({
        message: error.message
      })

    }

  }

};

module.exports = UserController;
